test/Constraints/sr12309.swift

```diff
@@ -0,0 +1,7 @@
+// RUN: %target-typecheck-verify-swift
+
+struct Foo {
+  var bar: Int
+}
+
+let f = Foo(bar: nil!) // expected-error {{'nil' literal cannot be force unwrapped}}
```

Comment:
Could you please move this test to `test/Constraints/optional.swift` and add a couple more tests too e.g. `_ = nil?` and `_ = (nil)` to make sure some easy cases are handled properly?

Reply:
For some reason when I added the tests in optional.swift the tests pass even when the shouldn't. Is there's something I'm doing wrong?

Reply:
I tried adding this and test fails:

```swift
// SR-12309 - Force unwrapping 'nil' compiles without warning
func sr_12309() {
  struct S {
    var foo: Int
  }

  _ = S(foo: nil!)
  _ = nil!
  _ = nil?
}
```

Reply:
Both of these cases: `_ = nil!` and `_ = S(foo: nil!)` are passing tests. I also handled _ = nil? like this one [optional.swift#L370](https://github.com/apple/swift/blob/b5cb81bd0a721d7c94b9243f86a4ffc0378546e0/test/Constraints/optional.swift#L370)
However tho this test _ = (nil) I can't handle properly.

Reply:
To handle `_ = (nil)` you'd have to move your code as I suggested and strip parens from `parentExpr` via `getSemanticsProvidingExpr()` it would degrade into `_ = nil` case.

Reply:
I can successfully check `_ = (nil)`  as you said in `visitNilLiteralExpr` by `isa<NilLiteralExpr>(parentExpr->getSemanticsProvidingExpr())` 

However tho... doing this cause failure in a lot of other tests

Reply:
When I moved my code to `visitNilLiteralExpr` instead a lot of other test cases failed as well

Reply:
Let's take a moment and consider what `parentExpr` is going to be here - it's a parent of `nil` literal, which means it could be either a `!` - `ForceValueExpr`, `?` - `OptionalEvaluationExpr` or it could be source (right-hand side of the assignment) so parent is going to be `AssignExpr` etc.

Reply:
Thank you for clarification, so doing this 
```
      if (isa<ForceValueExpr>(parentExpr)) {
        DE.diagnose(expr->getLoc(), diag::cannot_force_unwrap_nil_literal);
        return Type();
      }

      if (isa<OptionalEvaluationExpr>(parentExpr)) {
        DE.diagnose(expr->getLoc(), diag::unresolved_nil_literal);
        return Type();
      }

      if (isa<NilLiteralExpr>(parentExpr->getSemanticsProvidingExpr())) {
        DE.diagnose(expr->getLoc(), diag::unresolved_nil_literal);
        return Type();
      }
``` 
Passes all tests in the test file for this problem. However, it's failing other test files.

Reply:
When I also added it inside this [if statement](https://github.com/apple/swift/blob/448a14e15f74516f1d832ddefcc995aecb5b5edd/lib/Sema/CSGen.cpp#L1162) the diagnostic message doesn't get called.

Reply:
Maybe `getSemanticsProvidingExpr` is too heavy handed in this case, I'd suggest to do this:

```
auto *parentExpr = CS.getParentExpr(expr);
if (auto *PE = dyn_cast<ParenExpr>(parentExpr))
  parentExpr = PE->getSubExpr();
```

This way we'd only look through parens.